Show movies

CoderKiwi opened the pull request:
Closes #3

CoderKiwi committed:
Add nav bar with app title and 'Discover' tab

- import bootstrap-vue

CoderKiwi committed:
Add example movie tile rows to Discover page

- fixed some " and ' differences

CoderKiwi committed:
Add MediaTile component to show image with caption

CoderKiwi committed:
Add MediaHorizontalList showing horizontally-scrollable tiles to Home

CoderKiwi committed:
Hide scrolling bar in MediaHorizontalList

CoderKiwi committed:
Fix spacing between tiles, refactor so Home passes in mediaTiles to list

CoderKiwi committed:
Remove unnecessary 'Discover' tab in navbar

CoderKiwi committed:
Fix build by asserting not null in prop

CoderKiwi committed:
Add get popular movies from Trakt (without image thumbs)

- Use Movie and Ids classes to decouple views from API
- Use env var for Trakt client ID
- Add axios dependency

CoderKiwi committed:
Add trending movies list to home page

CoderKiwi committed:
Add most anticipated and box office hit movie lists to home page

CoderKiwi committed:
Refactor TraktApi to reuse a method to convert JSON to ts objects

CoderKiwi committed:
Get thumbnails of media

CoderKiwi committed:
Cleanup

- cleanup API classes
- fixing tslint errors

CoderKiwi committed:
Add showing no-image during loading tiles

- Change APIs to singletons
- Remove unnecessary HelloWorld component
- Skip unit tests

CoderKiwi commented on src/models/Movie.ts:
```diff
@@ -0,0 +1,33 @@
+        this._year = newYear;
+        this._ids = newIds;
+        new FanartApi().getMovieThumb(this._ids.imdb)
+            .then((result) => this._thumb = result);
```
Need to catch the potential error here

alextompkins replied:
This shouldn't be done within a model class - violates single responsibility principle as the class may need to be changed to reflect different method of retrieving thumbnail (e.g. caching etc. in future). 

Surely this property might be accessed before the request completes? In which case only the blank value will be used. 

Will a Movie object always require a thumbnail URL? What if this object is constructed when deserialized e.g. from JSON?

Potential solution: provide a setter for this property and assign it when needed. Provide FanartApi via dependency injection in the class that does this

CoderKiwi committed:
Refactor home page to avoid magic strings

CoderKiwi committed:
Cleanup tslint errors

CoderKiwi committed:
Cleanup tslint errors

CoderKiwi committed:
Merge remote-tracking branch 'origin/3-show-movies' into 3-show-movies

CoderKiwi commented on src/components/MediaTile.vue:
```diff
@@ -0,0 +1,30 @@
+        <b-img-lazy class="image" :src="imgSourceUri" :alt="title+' Thumbnail'" fluid rounded/>
+        <div>
+            {{ title }} ({{ year }})
+            <!--            todo make the text slide across when too long for image-->
```
Complete this. Perhaps in a different PR?

CoderKiwi replied:
Will not fix right now

CoderKiwi commented on src/models/Movie.ts:
```diff
@@ -0,0 +1,33 @@
+        this._title = newTitle;
+        this._year = newYear;
+        this._ids = newIds;
+        FanartApi.instance.getMovieThumb(this._ids.imdb)
```
This line may be overwriting the default `no-image.png` filepath in the `MediaTile`

CoderKiwi commented on src/services/FanartApi.ts:
```diff
@@ -0,0 +1,36 @@
+        }).then((response) => {
+            return response.data.moviethumb[0].url as string;
+        });
+        //     .catch(error => {
```
?

CoderKiwi commented on src/services/TraktApi.ts:
```diff
@@ -0,0 +1,65 @@
+import Movie from '@/models/Movie';
+import Ids from '@/models/Ids';
+
+// todo need to handle errors
```
Clean up

alextompkins commented on src/models/Movie.ts:
```diff
@@ -0,0 +1,33 @@
+    private readonly _title: string;
+    private readonly _year: number;
+    private readonly _ids: Ids;
+    private _thumb: string = '';
```
Change property name to show this is just a URL, not the image itself

alextompkins commented on src/components/MediaTile.vue:
```diff
@@ -0,0 +1,30 @@
+<template>
+    <div class="MediaTile">
+        <b-img class="image" :src="imgSourceUri" :alt="title+' Thumbnail'" fluid rounded/>
```
```suggestion
        <b-img class="image" :src="imgSourceUri" :alt="title+' Thumbnail'" fluid rounded></b-img>
```
Apparently "empty tag doesn't work in some browsers".

CoderKiwi replied:
Interesting...

alextompkins commented on src/components/MediaHorizontalList.vue:
```diff
@@ -0,0 +1,38 @@
+<template>
+    <div>
```
Is this `<div>` necessary?

alextompkins commented on src/components/MediaHorizontalList.vue:
```diff
@@ -0,0 +1,38 @@
+<template>
+    <div>
+        <b-container class="container">
+            <div v-for="mediaObj in mediaObjs" :key='mediaObj.ids.imdb'>
```
Is this `<div>` necessary? Put v-for on MediaTile instead

CoderKiwi replied:
No, turns out the `<div>` isn't necessary. I'll remove that. 

I'm not sure I understand the `v-for` suggestion. Could you elaborate? I don't think I could remove that `<div>` on line 4 if that's what you're asking.

alextompkins replied:
To clarify, the `v-for` attribute can be on *any* element, not just divs. Hence you could place it (along with the `:key` attribute) on the MediaTile element instead and have exactly the same effect. The `<div>` would then be unnecessary.

alextompkins commented on src/components/MediaHorizontalList.vue:
```diff
@@ -0,0 +1,38 @@
+        overflow-x: auto;
+        min-width: 100%;
+    }
+    ::-webkit-scrollbar {
```
This works for Chome and Safari (webkit-based), for MS Edge and Firefox do this:
Edge: `-ms-overflow-style: -ms-autohiding-scrollbar;`
Firefox: `scrollbar-width: none;`
on .container

CoderKiwi replied:
Nice catch -- I only tested with Chrome and Safari. 

However with this code:
```suggestion
    .container {
        display: flex;
        overflow-x: auto;
        min-width: 100%;
        -ms-overflow-style: -ms-autohiding-scrollbar; /* Edge */
        scrollbar-width: none; /* Firefox */
    }

    ::-webkit-scrollbar {
        display: none;
    }
```

WebStorm complains: 
>Unknown CSS property 'scollbar-width'

Did you test it with Firefox?

alextompkins replied:
On further inspection, `scrollbar-width` "is part of the CSS Working Group's Scrollbars Module Level 1 draft, which is still a work in progress". It has experimental support in Firefox, but no other browsers:
https://caniuse.com/#search=scrollbar-width

Hence the complaint.

alextompkins commented on src/views/Home.vue:
```diff
@@ -1,18 +1,58 @@
+        private boxOfficeMovies: Movie[] = [];
+
+        private async mounted() {
+            const trakt = new TraktApi(); // todo
```
Deliver a TraktApi to this component instead via dependency injection. This will allow for mocking and achieve inversion of control

CoderKiwi replied:
> Deliver a TraktApi to this component instead via dependency injection. This will allow for mocking and achieve inversion of control

[InveresifyJS](https://github.com/inversify/InversifyJS) looks like a good option.

CoderKiwi replied:
Will need to update [Decision Decisions](https://github.com/CoderKiwi/Volo/wiki/Design-decisions)

CoderKiwi replied:
This was worked on as part of #5

alextompkins commented on src/views/Home.vue:
```diff
@@ -1,18 +1,58 @@
+
+        private async mounted() {
+            const trakt = new TraktApi(); // todo
+            await Promise.all([
```
Why wait for all of these to complete? There shouldn't be anything happening after the mounted lifecycle hook so don't even bother making this an async method, instead remove the `Promise.all()`

alextompkins reviewed (changes requested):
### Singletons & Dependency Injected APIs
The singletons for axios instances are fine being locked away as statics within the Api services. However, the Api objects should be delivered via dependency injection. I need to do some more research into what options are available for dependency injection in Vue though.

CoderKiwi committed:
PR review changes

- Remove Promise.all when loading Home page media tiles
- Remove unneeded div
- Remove unneeded inline comments

CoderKiwi committed:
Refactor TraktApi method to use .map

Co-Authored-By: Alex Tompkins <alextompkins3@gmail.com>

CoderKiwi committed:
Fix bad GitHub PR review commit

CoderKiwi committed:
Refactor TraktApi to use .map

CoderKiwi commented:
> ### Singletons & Dependency Injected APIs
> The singletons for axios instances are fine being locked away as statics within the Api services. However, the Api objects should be delivered via dependency injection. I need to do some more research into what options are available for dependency injection in Vue though.

@alextompkins [this article](https://markus.oberlehner.net/blog/dependency-injection-in-vue-applications/#plugins) seems pretty good. Not sure which one's best between plugins or provide/inject. Thoughts?

CoderKiwi committed:
Removed unneeded div in MediaHorizontalList

CoderKiwi committed:
Add Edge, Firefox CSS to disable horizontal scrollbars in horz lists

CoderKiwi committed:
Change FanartApi to try/catch errors

- Change Vue components to not modify props (anti-pattern)
- Remove marquee todo which will not be fixed as part of this issue
- Remove lazy loading MediaTile images

CoderKiwi commented on src/services/FanartApi.ts:
```diff
@@ -0,0 +1,26 @@
+            params: {
+                api_key: this.projectKey,
+            },
+        }).then(function(result) {
```
Done

CoderKiwi commented on src/services/FanartApi.ts:
```diff
@@ -0,0 +1,26 @@
+            return result.data.moviethumb[0].url as string;
+        }).catch(function(error) {
+            console.log();
+            return ''; // todo fix
```
> Change to handle two distinct cases:
> 
> 1. No thumbnail exists for this movie --> return null and handle in caller (add null to contract)

Instead of returning `null` I'm simply returning the URI to the `no-image.png`

CoderKiwi committed:
Apply suggestions from code review

Co-Authored-By: Alex Tompkins <alextompkins3@gmail.com>

CoderKiwi commented on src/services/FanartApi.ts:
```diff
@@ -0,0 +1,37 @@
+import axios, {AxiosInstance} from 'axios';
+
+export default class FanartApi {
```
Should make this implement an interface to make testing via mocking easier, etc.

CoderKiwi replied:
Same with `TraktApi`

CoderKiwi commented:
The `no-image.png` showing is fine, however it results in thumbs that do not appear 'clickable'. Perhaps a border will help?

CoderKiwi committed:
Add API service interfaces

CoderKiwi commented:
<img width="682" alt="Screen Shot 2019-11-13 at 11 17 42 PM" src="https://user-images.githubusercontent.com/33564041/68754560-ce433600-066b-11ea-9db0-1763a7269157.png">

Notice the caption is lower on the right tile? The `no-image.png` is actually larger than the other movie thumbnails. This should be fixed so every image is the same size and aspect ratio.

CoderKiwi commented:
<img width="195" alt="Screen Shot 2019-11-14 at 1 12 18 AM" src="https://user-images.githubusercontent.com/33564041/68762819-d4411300-067b-11ea-9cec-f44ad9b6b782.png">

An anticipated movie which does not have a release date yet will show an empty `()`

CoderKiwi committed:
Add About modal which links to API resources Trakt, Fanart

CoderKiwi committed:
Fix showing empty '()' if no year given in media tile

CoderKiwi committed:
Remove unnecessary console.log()

CoderKiwi committed:
Merge branch 'master' into 3-show-movies

alextompkins committed:
Fix hide-footer prop type error

Took 59 minutes

alextompkins commented:
## Review: Story Outcomes
![image](https://user-images.githubusercontent.com/18250951/68992103-ad791b80-08cb-11ea-9059-9b3aba720193.png)
1. The movies should be shown as horizontal lists of tiles
**Passed**
2. Each horizontal list should have a title describing the list (most anticipated, trending, etc.)
**Passed**
3. The top ten of each category is shown
**Passed**
4. The background of the tile should be an appropriate image of the movie
**Passed**
5. The title of the movie should be shown
**Passed**
6. The year of the movie should be shown
**Passed**
